### PR TITLE
Promql cpp functions

### PR DESCRIPTION
**Pull request by cherep58**
## Summary

Implement C++ pushdown optimizations for PromQL functions at the storage layer. When the query engine provides `SelectHints.Func`, the storage can now reduce data before it reaches the Go PromQL engine — returning only the samples needed for the specific function instead of the full series.

### New decorator iterators

- **Aggregation:** `MinOverTimeIterator`, `MaxOverTimeIterator`, `LastOverTimeIterator`, `SumOverTimeIterator` (via `OverTimeFuncIterator` template with pluggable handlers)
- **Counter/rate family:** `RateIterator` (rate, increase), `IRateIterator` (irate, idelta)
- **Transform:** `DeltaIterator`, `ChangesIterator`, `ResetsIterator`
- **Downsampling:** `DownsamplingDecodeIterator` — interval-based sample reduction

### Refactored decode iterator infrastructure

- CRTP-based `DecodeIteratorTrait` with `seek()`, `seek_to()`, `invalidate()`, `set()` methods
- `SeekResult` enum for composable seek logic (`kUpdateSample`, `kNext`, `kStop`, `kUpdateSampleNextAndStop`)
- `UniversalDecodeIterator` extended with `seek()`, `seek_to()`, `invalidate()`, `set()` that dispatch through `std::visit`
- Outer `DecodeIterator` variant wrapping all iterator types for the Go binding layer

### Function dispatch

- gperf-generated perfect hash (`FunctionNamesHash`) for O(1) function name lookup
- `create_decode_iterator()` dispatches on `SelectHints.Func` to construct the appropriate iterator

### Go bridge changes

- `SelectHints` passed through cppbridge via ABI-compatible `GenericSelectHints<Go::String, Go::SliceView>`
- `downsamplingMs` parameter added to `Query()` and `ChunkRecoder` paths
- `DataStorageSerializedDataIterator` control block simplified: direct `Timestamp()`/`Value()` accessors replacing raw field access
- Namespace change: `entrypoint::head` → `entrypoint::series_data`

### Testing

- Parametric C++ unit tests for each new iterator (boundary conditions, StaleNaN handling, time interval filtering, reset scenarios)
- Go integration tests for all dispatched PromQL functions (downsampling, min/max/last/sum_over_time, rate, increase, changes, delta, irate, idelta, resets)

### CODE REVIEW

File: pp/entrypoint/types/aggregation_iterator.h [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    case promql_function_name_hash("sum_over_time"):
+      return DecodeIterator(std::in_place_type<DecodeIterator::SumOverTimeIterator>, select_hints.interval);
+
+    case promql_function_name_hash("delta"):
+      return DecodeIterator(std::in_place_type<DecodeIterator::DeltaIterator>, select_hints.interval);
+
+    case promql_function_name_hash("resets"):
+      return DecodeIterator(std::in_place_type<DecodeIterator::ResetsIterator>, select_hints.interval);
+
+    default:
```
Comment (vporoshok) [inline pp/entrypoint/types/aggregation_iterator.h:247]: **Bug: `"changes"` is not dispatched.**

`ChangesIterator` is implemented, added to `IteratorVariant`, included in the header, and has C++ tests — but there is no `case promql_function_name_hash("changes"):` here. The function falls through to `default` and returns `UniversalDecodeIterator` (no optimization).

The Go test `TestChangesFunc` confirms this — it expects all original samples as if no pushdown happened.

Either:
- Add the missing case, or
- Remove `ChangesIterator` from `IteratorVariant` and the includes if it's intentionally deferred — otherwise it's dead code inflating the variant size.

File: pp/series_data/decoder/decorator/over_time_func_iterator.h [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    const auto result = *this;
+    ++*this;
+    return result;
+  }
+
+ protected:
+  UniversalDecodeIterator iterator_;
+  PromPP::Primitives::TimeInterval interval_;
+
+  void find_element() {
```
Comment (vporoshok) [inline pp/series_data/decoder/decorator/over_time_func_iterator.h:45]: **Bug: aggregation over `[Start, End]` breaks range queries.**

`find_element()` aggregates ALL non-stale samples in `[interval_.min, interval_.max]` into a single result. For a range query with step, the PromQL engine evaluates the function at each step point within its own window `[t-Range, t]`. If the storage returns a single aggregated sample for the entire series, the engine will use it for all step points — producing incorrect results when the aggregate differs between windows.

Example: `min_over_time(metric[5m])` with step=1m — the minimum in `[t-5m, t]` can differ for different `t`, but the storage returns one global minimum over the full `[Start, End]`.

Affects `MinOverTimeIterator`, `MaxOverTimeIterator`, `SumOverTimeIterator`, `LastOverTimeIterator`.

File: pp/entrypoint/series_data/decode_iterator.h [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+PROMPP_ALWAYS_INLINE DecodeIterator create_decode_iterator(const PromPP::Prometheus::SelectHints& select_hints, PromPP::Primitives::Timestamp downsampling_ms) {
+  if (downsampling_ms != ::series_data::decoder::decorator::kNoDownsampling) [[unlikely]] {
+    return DecodeIterator(std::in_place_type<DecodeIterator::DownsamplingIterator>, downsampling_ms);
+  }
+  if (select_hints.func.empty()) [[likely]] {
+    return DecodeIterator(std::in_place_type<DecodeIterator::UniversalDecodeIterator>);
+  }
+
+  switch (promql_function_name_hash(select_hints.func)) {
```
Comment (vporoshok) [inline pp/entrypoint/series_data/decode_iterator.h]: **Design: gperf hash without string verification.**

The switch dispatches on `promql_function_name_hash()` which is a perfect hash only among the 81 known function names. An arbitrary string not in the gperf table could produce a hash value matching one of the handled cases, silently applying the wrong optimization.

I verified that among the current 81 PromQL function names no unhandled one collides with a handled case — so this is safe **today**. But it's fragile against future additions. Consider adding a string equality check after the hash match, or a compile-time/test-time collision check.

File: pp/go/storage/block/writer.go [hunk 4]
```diff
@@ -121,6 +110,25 @@ func (w *Writer[TShard]) createWriters(sd TShard) (blockWriters, error) {
...
+	})
+
+	return newBlockWriter(
+		dataDir,
+		w.maxBlockChunkSegmentSize,
+		NewIndexWriter(lss),
+		chunkIterator,
+	)
+}
 // recodeAndWriteChunks recodes and writes chunks for the shard.
```
Comment (vporoshok) [inline pp/go/storage/block/writer.go:167]: Nit: missing blank line between `createWriter` closing brace and this comment. Go convention requires a blank line between top-level declarations.

**Review by vporoshok — CHANGES_REQUESTED** [review]
Overall solid work — well-structured decorator iterators, good CRTP-based seek infrastructure, thorough test coverage. Two bugs and a few minor items below.

File: pp/entrypoint/series_data_data_storage.cpp [hunk 5]
```diff
@@ -123,18 +123,20 @@ extern "C" void prompp_series_data_data_storage_query_v2(void* args, void* res)
...
     QueryStatus status{};
-    entrypoint::head::SerializedDataPtr* serialized_data{};
+    entrypoint::series_data::SerializedDataPtr* serialized_data{};
   };
 
   const auto in = static_cast<Arguments*>(args);
   const auto out = static_cast<Result*>(res);
 
-  RangeQuerierWithArgumentsWrapperV2 querier(*in->data_storage, in->query, out->serialized_data);
+  RangeQuerierWithArgumentsWrapperV2 querier(*in->data_storage, in->query, *in->hints, out->serialized_data, in->downsampling_ms);
```
Comment (gshigin) [inline pp/entrypoint/series_data_data_storage.cpp]: Do we always know that hints are valid? Go's `SelectHints` states that this parameter is optional. Even in our code (pp_api.go) : `seriesSet := q.Select(ctx, false, nil, matchers...)`, where hint is nil

Comment (cherep58) [reply pp/entrypoint/series_data_data_storage.cpp]: Hints is mandatory parameter for range query in cpp. I will research go code and fix nil hints, thanks)

File: pp/go/storage/querier/series_bench_test.go [hunk 6]
```diff
@@ -28,7 +28,7 @@ func iterateSeriesSet(seriesSet storage.SeriesSet) {
 	}
 }
 
-func queryOpt(t testing.TB, lss *shard.LSS, ds *shard.DataStorage, start, end int64, matchers ...model.LabelMatcher) *querier.SeriesSet {
+func queryOpt(t testing.TB, lss *shard.LSS, ds *shard.DataStorage, start, end, downsamplingMs int64, matchers ...model.LabelMatcher) *querier.SeriesSet {
```
Comment (gshigin) [inline pp/go/storage/querier/series_bench_test.go:32]: This change breaks other test calls in `merge_series_set_test.go`. Either fix other calls, or remove downsamplingMs, because here it's only used with `NoDownsampling`

File: pp/entrypoint/series_data/serialization.h [hunk 7]
```diff
@@ -1,28 +1,42 @@
 #pragma once
 
+#include "entrypoint/series_data/decode_iterator.h"
+#include "primitives/go_slice.h"
+#include "primitives/primitives.h"
+#include "prometheus/query.h"
 #include "series_data/serialization/serialized_data.h"
 
-namespace entrypoint::head {
+namespace entrypoint::series_data {
```
Comment (gshigin) [inline pp/entrypoint/series_data/serialization.h]: namespace is changed but file is still at `entrypoint/head.h`. Consider moving it or renaming it back

File: pp/entrypoint/series_data_data_storage.h [hunk 8]
```diff
@@ -98,12 +98,13 @@ void prompp_series_data_data_storage_allocated_memory(void* args, void* res);
  * @param args {
  *     dataStorage    uintptr          // pointer to constructed data storage
  *     query          DataStorageQuery // query
+ *     downsamplingMs int64            // downsampling interval in milliseconds (0 - downsampling is disabled)
```
Comment (gshigin) [inline pp/entrypoint/series_data_data_storage.h]: Add hints to args parameters.

Also, it's possible to pass to this function valid downsampling timestamp and valid hint at the same time. And implementation gives preference to construct `DownsamplingIterator` in such case. Any sort of contract validation will do: comment here that its not possible to create such iterator, or `assert` somewhere in `create_decode_iterator()`, or anything else

File: pp/prometheus/query.h [hunk 9]
```diff
@@ -1,23 +1,58 @@
...
 #include "label_matcher.h"
+#include "primitives/primitives.h"
 
 namespace PromPP::Prometheus {
 
-struct ReadHints {
-  std::string func{};
-  std::vector<std::string> grouping;
+template <class String, template <class> class Slice>
+struct GenericSelectHints {
```
Comment (gshigin) [inline pp/prometheus/query.h:11]: Do we need this as heavy template class? I've seen extensive use of `GoSelectHints`, but not as much for `SelectHints = GenericSelectHints<std::string, std::vector>;`.  `Query` class has `hints` member, but nobody reads it or writes it. 

We also have another struct with exact same name in different place: `entrypoint::series_data::SelectHints`. 

If we have heavy use of `GoSelectHints` and `entrypoint::series_data::SelectHints`, it's better to make them concrete types with clear relations. Or at least fix this naming ambiguity

File: pp/series_data/decoder/decorator/changes_iterator.h [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    if (!has_value) [[unlikely]] {
+      iterator_.invalidate_sample();
+    }
+  }
+
+  PROMPP_ALWAYS_INLINE void seek_to_next_sample() {
+    bool has_changes{};
+
+    iterator_.seek<SeekKind::kAll>([&has_changes, this](PromPP::Primitives::Timestamp timestamp, double value) {
+      using BareBones::Encoding::Gorilla::isstalenan;
```
Comment (gshigin) [inline pp/series_data/decoder/decorator/changes_iterator.h:69]: Not used. Does `ChangesIterator `intentionally ignores StaleNaNs?

Comment (cherep58) [reply pp/series_data/decoder/decorator/changes_iterator.h:69]: `ChangesIterator` don't ignores StaleNaNs. `ChangesIterator` handle all values

File: pp/prometheus/query.h (hunk 9 above)
Comment (cherep58) [reply pp/prometheus/query.h:11]: Go primitive types should be used only in entrypoint. Other code should operate template classes or c++ primitives (span etc).

GoSelectHints and SelectHints in entrypoint have different names, we haven't naming ambiguity.

File: pp/series_data/decoder/decorator/window_function_iterator.h [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+
+  [[nodiscard]] PROMPP_ALWAYS_INLINE TimeInterval advance_interval() const noexcept {
+    auto interval = iterator_.interval();
+    if (interval.difference() == parameters_->step) {
+      interval.min = interval.max;
+
+      const auto diff = parameters_->range - parameters_->step;
+      interval.max += (diff == 0) ? parameters_->step : diff;
+    } else {
+      interval.min = std::exchange(interval.max, next_interval_boundary(interval.min));
```
Comment (vporoshok) [inline pp/series_data/decoder/decorator/window_function_iterator.h]: **Bug (critical): `advance_interval()` produces invalid intervals when `w > 2s`**

Trace for `w = 9, s = 4, d = 1` (e.g. `max_over_time(m[9m])` with 4m step):

```
1st interval:  [c, c+4]       len 4 = s         ✓
2nd:           diff == step → [c+4, c+9]  len 5 = w−s
3rd:           diff ≠ step → min = c+9,
               max = next_boundary(c+4) = c+8
               → [c+9, c+8]   min > max!  INVALID
```

When `w − s > s`, the second interval overshoots the next grid boundary. `next_interval_boundary(old.min)` returns a value less than the current `max`, producing an inverted interval. The `advance()` loop will either spin forever or produce UB on the unsigned difference.

Root cause: `advance_interval` alternates between adding `step` and `w − s`, but assumes `w − s ≤ s`. This breaks for `w > 2s`.

Per ADR-001, when `w > s` and `d ≠ 0`, the boundary grid should alternate at `d` and `s − d` intervals (not `s` and `w − s`). For `w > 2s` the grid still works: boundaries at `a + n·s` and `a + n·s − d` with sub-interval lengths always `d` and `s − d`.

File: pp/series_data/decoder/decorator/window_function_iterator.h [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+  [[nodiscard]] PROMPP_ALWAYS_INLINE Timestamp next_interval_boundary(Timestamp start) const noexcept {
+    return next_interval_boundary(start, parameters_->step, parameters_->range);
+  }
+
+  [[nodiscard]] PROMPP_ALWAYS_INLINE static Timestamp next_interval_boundary(Timestamp start, Timestamp step_ms, Timestamp range_ms) noexcept {
+    if (range_ms <= step_ms) [[likely]] {
+      return start + range_ms;
+    }
+
+    return start + range_ms - (range_ms - step_ms);
```
Comment (vporoshok) [inline pp/series_data/decoder/decorator/window_function_iterator.h]: **Bug (critical): `next_interval_boundary()` produces wrong grid for `w < s`**

When `w ≤ s`, this returns `start + w`. Trace for `w = 2, s = 5`:

```
Go windows: (c, c+2], (c+5, c+7], (c+10, c+12], ...  (gaps of 3)

Code produces:
[c, c+2]     ✓ (first window)
[c+2, c+2]   len 0 → skip
[c+2, c+4]   ← WRONG: not aligned to any Go window!
[c+4, c+4]   len 0 → skip
[c+4, c+6]   ← partially in gap (c+4,c+5] + partially in window (c+5,c+6]
```

The function creates a grid with spacing `w` regardless of `s`. But per ADR-001 case 3 (`w < s`), boundaries should be at `a + n·s` and `a + n·s − w`, giving alternating sub-intervals of length `s − w` (gap, skip) and `w` (actual window). The correct grid spacing is driven by `s`, not `w`.

Suggestion: `next_interval_boundary` should always produce boundaries from the grid `{a + n·s, a + n·s − d}` where `d = w % s`.

File: pp/series_data/decoder/decorator/sum_over_time.h [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+
+class SumOfElements {
+ public:
+  PROMPP_ALWAYS_INLINE SeekResult operator()(PromPP::Primitives::Timestamp timestamp, double value) noexcept {
+    if (BareBones::Encoding::Gorilla::isstalenan(sum_.value)) [[unlikely]] {
+      sum_.value = 0.0;
+    }
+
+    kahan_sum_inc(value, sum_.value, c_);
+    sum_.timestamp = timestamp;
```
Comment (vporoshok) [inline pp/series_data/decoder/decorator/sum_over_time.h:30]: **Moderate: possible double-count at interval boundaries**

Consecutive sub-intervals share a boundary: new interval's `min` = previous interval's `max`. `OverTimeFuncIterator::find_element()` calls `seek_to(interval_.min)` which positions at the first sample `≥ min`.

If a sample has a timestamp exactly equal to the shared boundary, it would be included in **both** adjacent sub-intervals. For `max`/`min` this is idempotent, but for `sum_over_time` it causes a double-count.

Please verify: does `seek_to` position at `> min` (exclusive) or `≥ min` (inclusive)?

File: pp/entrypoint/types/aggregation_iterator.h [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+    case kDelta:
+      return DecodeIterator(std::in_place_type<DecodeIterator::DeltaIterator>, select_hints.function_parameters);
+
+    case kResets:
+      return DecodeIterator(std::in_place_type<DecodeIterator::ResetsIterator>, select_hints.function_parameters);
+
+    case kChanges:
+      return DecodeIterator(std::in_place_type<DecodeIterator::ChangesIterator>, select_hints.function_parameters);
+
+    default:
```
Comment (vporoshok) [inline pp/entrypoint/types/aggregation_iterator.h:247]: **Design: `count_over_time` not dispatched**

ADR-001 lists `count_over_time` in the per-function output contract. It's not in the `WindowFunction` enum and falls through to the `default` branch here (universal iterator). This is functionally correct but misses the pushdown optimization.

Considering it only needs a count per sub-interval (similar to `sum_over_time`), it would be a small addition.

**Review by vporoshok — CHANGES_REQUESTED** [review]
## Review: WindowFunctionIterator sub-interval logic vs ADR-001

Analyzed the sub-interval splitting algorithm in `WindowFunctionIterator` against [ADR-001](https://fox.flant.com/team/observability/prompp-wiki/-/blob/adr/range-function-iteration/adr/001-range-function-iteration-algorithm.md) (range function iteration algorithm). The code works correctly for the most common case (`s < w < 2s`), but has critical issues in two other cases.

### Summary of findings

| # | Severity | Finding |
|---|---|---|
| 1 | **Bug/Critical** | `advance_interval()` breaks for `w > 2s` — produces invalid interval (min > max), potential infinite loop |
| 2 | **Bug/Critical** | Incorrect boundaries for `w < s` — aggregation captures samples from gaps between windows |
| 3 | **Bug/Moderate** | Possible double-count for `sum_over_time` at interval boundaries (depends on `seek_to` inclusive/exclusive semantics) |
| 4 | **Design** | `count_over_time` missing from dispatch (falls back to universal iterator — correct but no pushdown) |
| 5 | **Design** | `sum_over_time` result timestamp is last raw sample's ts, not sub-interval end |
| 6 | **Observation** | min/max sub-interval collapsing optimization from ADR not implemented (correct but sub-optimal) |

File: pp/series_data/decoder/decorator/sum_over_time.h (hunk 13 above)
Comment (cherep58) [reply pp/series_data/decoder/decorator/sum_over_time.h:30]: No, code has valid behavior. I added tests in `SumOverTimeWindowFunctionIteratorFixture`

File: pp/entrypoint/series_data_data_storage.cpp (hunk 5 above)
Comment (cherep58) [reply pp/entrypoint/series_data_data_storage.cpp]: I added a nullptr hints validation

File: pp/go/cppbridge/head.go [hunk 15]
```diff
@@ -246,39 +259,83 @@ func (sd *DataStorageSerializedData) Next() (uint32, uint32) {
 	return seriesDataSerializedDataNext(sd.serializedData)
```
Comment (vporoshok) [inline pp/go/cppbridge/head.go]: **Bug: incorrect `runtime.KeepAlive`.** The `KeepAlive` is placed *before* the cgo call (a no-op for its purpose) and keeps `sd.ds`, while the finalizer that frees `sd.serializedData` is attached to `sd` itself. The GC may reclaim `sd` and run `seriesDataSerializedDataDtor` during the cgo call → use-after-free. It should be:

```go
seriesID, chunkRef := seriesDataSerializedDataNext(sd.serializedData)
runtime.KeepAlive(sd)
return seriesID, chunkRef
```

The same pattern is worth double-checking in the iterator `Ctor`/`Reset` bridge calls — wrap them with `runtime.KeepAlive(serializedData)`.

File: storage/interface.go [hunk 16]
```diff
@@ -214,10 +214,19 @@ type SelectHints struct {
...
 	// DisableTrimming allows to disable trimming of matching series chunks based on query Start and End time.
 	// When disabled, the result may contain samples outside the queried time range but Select() performances
 	// may be improved.
 	DisableTrimming bool
+
+	By bool // Indicate whether it is without or by.
+
+	// IsSubquery indicates whether the query is a subquery.
+	// For subquery, we need to disable the optimization of the query
+	IsSubquery bool
```
Comment (vporoshok) [inline storage/interface.go:229]: **Fragile ABI contract.** `GoSelectHints` is reinterpret-cast directly from `storage.SelectHints`; the layout only matches thanks to the manual field reordering here (`By`, `LookbackDelta`, `IsSubquery`). Any future change to `SelectHints` (e.g. an upstream rebase) will silently break the C++ parsing. Please add a safety net:
- add `Sizeof_GoSelectHints` to `go_constants` and a `static_assert(sizeof(GoSelectHints) == Sizeof_GoSelectHints)` in `go_constants.cpp` (as already done for the other structs);
- a Go test using `unsafe.Sizeof`/`unsafe.Offsetof` cross-checking against those constants;
- a warning comment on both structs that the field order is coupled.

File: pp/series_data/decoder/decorator/downsampling_decode_iterator.h [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+        return;
+      }
+    }
+  }
+
+ private:
+  DecodeIterator iterator_;
+  Timestamp interval_;
+
+  PROMPP_ALWAYS_INLINE static Timestamp round_up_to_step(Timestamp timestamp, Timestamp step) noexcept {
```
Comment (vporoshok) [inline pp/series_data/decoder/decorator/downsampling_decode_iterator.h:57]: **Incorrect for negative timestamps.** `result % step` yields a negative remainder in C++, so the rounding is wrong for `timestamp < 0`. Suggest floor-based rounding (`ceil(a/b) = -floor(-a/b)`), reusing a shared `floor_div` — it is currently duplicated in `window_boundary_calculator.h`, so it would be good to hoist it into `bare_bones/algorithm.h`:

```cpp
template <class T>
constexpr T floor_div(T numerator, T denominator) noexcept {
  auto quotient = numerator / denominator;
  if (numerator % denominator < 0) {
    --quotient;
  }
  return quotient;
}
```

```cpp
PROMPP_ALWAYS_INLINE static Timestamp round_up_to_step(Timestamp timestamp, Timestamp step) noexcept {
  return -BareBones::floor_div(-timestamp, step) * step;
}
```

Checks out: `1000→1000`, `1001→2000`, `-1001→-1000`.

**Review by vporoshok — COMMENTED** [review]
A few points from reviewing the pushdown/optimization work. Great structure overall — CRTP seek infra, per-window `WindowFunctionIterator`, and the test coverage are solid. One correctness bug (KeepAlive), one hardening request (ABI contract), one edge-case fix (negative timestamps), plus two nits.

File: pp/go/storage/querier/querier.go [hunk 18]
```diff
@@ -305,12 +462,67 @@ func (q *Querier[TTask, TDataStorage, TLSS, TShard, THead]) selectRange(
...
+
+	if funcOptimizeMap[hints.Func]&queryOptimize != 0 && isNotWithpout(hints) {
+		return hints
+	}
+
+	return emptySelectHints
+}
+
+// isNotWithpout checks if the hints is not without by.
+func isNotWithpout(hints *storage.SelectHints) bool {
```
Comment (vporoshok) [inline pp/go/storage/querier/querier.go]: Nit: typo in the name `isNotWithpout` → `isNotWithout`.

File: pp/entrypoint/types/aggregation_iterator.h [hunk 19]
```diff
@@ -0,0 +1,248 @@
...
+
+      case Type::kChanges: {
+        return std::forward<Visitor>(visitor)(iterator_.changes);
+      }
+
+      case Type::kDelta: {
+        return std::forward<Visitor>(visitor)(iterator_.delta);
+      }
+
+      default: {
```
Comment (vporoshok) [inline pp/entrypoint/types/aggregation_iterator.h]: Nit: using `default:` for `kResets` means adding a new `Type` will silently miss a `visit` branch (no compiler warning). Consider an explicit `case Type::kResets:` and a `default:` that is unreachable/asserts.

File: pp/go/cppbridge/head.go (hunk 15 above)
Comment (cherep58) [reply pp/go/cppbridge/head.go]: In iterators ctor/reset functions serializedData are holded in other structures and need not to be keepalived

File: storage/interface.go (hunk 16 above)
Comment (cherep58) [reply storage/interface.go:229]: `Sizeof_GoSelectHints`, `unsafe.Sizeof` and comments will not solve the problem of fields reordering. We have unit tests in Go where SelectHints are passed and used by C++. Tests is only mechanism of struct layout validation, IMHO

File: pp/series_data/decoder/decorator/downsampling_decode_iterator.h (hunk 17 above)
Comment (cherep58) [reply pp/series_data/decoder/decorator/downsampling_decode_iterator.h:57]: Currently we don't fully support negative timestamps. If we want to support negative timestamps, we should refactor code base and write necessary unit tests

**Comment by cherep58** [comment]
- [ ] Create unit tests for entrypoint_types

**Review by vporoshok — APPROVED** [review]
